src/transformers/trainer_utils.py

```diff
@@ -46,7 +46,38 @@
     import tensorflow as tf
 
 
-def set_seed(seed: int):
+def seed_worker():
+    """
+    Helper function to set worker seed during Dataloader initialization.
+    """
+    worker_seed = torch.initial_seed() % 2**32
+    set_seed(worker_seed)
+
+
+def enable_determinism_for_distributed_training():
+    """
+    Helper function for reproducible behavior during distributed training. See
+    - https://pytorch.org/docs/stable/notes/randomness.html for pytorch
+    - https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism for tensorflow
+    """
+
+    if is_torch_available():
+        #  Enable PyTorch deterministic mode. This potentially requires either the environment
+        #  variable 'CUDA_LAUNCH_BLOCKING' or 'CUBLAS_WORKSPACE_CONFIG' to be set,
+        # depending on the CUDA version, so we set them both here
+        os.environ["CUDA_LAUNCH_BLOCKING"] = "1"
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":16:8"
+        torch.use_deterministic_algorithms(True)
+
+        # Enable CUDNN deterministic mode
+        torch.backends.cudnn.deterministic = True
+        torch.backends.cudnn.benchmark = False
+
+    if is_tf_available():
+        tf.config.experimental.enable_op_determinism()
+
+
+def set_seed(seed: int, enable_determinism: bool = True):
```

Comment:
```suggestion
def set_seed(seed: int, distributed_determinism: bool = False):
```
I like `distribitude_determinism` a bit better. Since this is a new addition, the default should be set to `False`. Although it does fix what one might consider a bug, so I'm not sure on this one. @LysandreJik do you have an opinion?